exec/java-exec/src/main/java/org/apache/drill/exec/planner/cost/DrillRelMdSelectivity.java

```diff
@@ -0,0 +1,219 @@
+    double sel = 1.0;
+    // determine which filters apply to the left vs right
+    RexNode leftPred = null;
+    RexNode rightPred = null;
+    JoinRelType joinType = rel.getJoinType();
+    final RexBuilder rexBuilder = rel.getCluster().getRexBuilder();
+    int[] adjustments = new int[rel.getRowType().getFieldCount()];
+
+    if (DrillRelOptUtil.guessRows(rel)) {
+      return super.getSelectivity(rel, predicate);
+    }
+
+    if (predicate != null) {
+      RexNode pred;
+      List<RexNode> leftFilters = new ArrayList<RexNode>();
+      List<RexNode> rightFilters = new ArrayList<RexNode>();
+      List<RexNode> joinFilters = new ArrayList<RexNode>();
+      List<RexNode> predList = RelOptUtil.conjunctions(predicate);
+
+      RelOptUtil.classifyFilters(
+          rel,
+          predList,
+          joinType,
+          joinType == JoinRelType.INNER,
+          !joinType.generatesNullsOnLeft(),
+          !joinType.generatesNullsOnRight(),
+          joinFilters,
+          leftFilters,
+          rightFilters);
+      leftPred =
+          RexUtil.composeConjunction(rexBuilder, leftFilters, true);
+      rightPred =
+          RexUtil.composeConjunction(rexBuilder, rightFilters, true);
+      for (RelNode child : rel.getInputs()) {
+        RexNode modifiedPred = null;
+
+        if (child == rel.getLeft()) {
+          pred = leftPred;
+        } else {
+          pred = rightPred;
+        }
+        if (pred != null) {
+          // convert the predicate to reference the types of the children
+          modifiedPred =
+              pred.accept(new RelOptUtil.RexInputConverter(
+              rexBuilder,
+              null,
+              child.getRowType().getFieldList(),
+              adjustments));
+        }
+        sel *= RelMetadataQuery.getSelectivity(child, modifiedPred);
+      }
+      sel *= RelMdUtil.guessSelectivity(
+          RexUtil.composeConjunction(rexBuilder, joinFilters, true));
+    }
+    return sel;
+  }
+
+  private Double getScanSelectivity(TableScan scan, RexNode predicate) {
+    DrillTable table = scan.getTable().unwrap(DrillTable.class);
+    if (table == null) {
+      table = scan.getTable().unwrap(DrillTranslatableTable.class).getDrillTable();
+    }
+    if (table == null || table.getStatsTable() == null) {
+      return super.getSelectivity(scan, predicate);
+    } else {
+      return getScanSelectivityInternal(table, predicate, scan.getRowType());
+    }
+  }
+
+  private Double getScanSelectivityInternal(DrillTable table, RexNode predicate,
+      RelDataType type) {
+    double sel = 1.0;
+
+    if ((predicate == null) || predicate.isAlwaysTrue()) {
+      return sel;
+    }
+
+    for (RexNode pred : RelOptUtil.conjunctions(predicate)) {
+      double orSel = 0;
+      for (RexNode orPred : RelOptUtil.disjunctions(pred)) {
+        //CALCITE guess
+        Double guess = RelMdUtil.guessSelectivity(pred);
+        if (orPred.isA(SqlKind.EQUALS)) {
```

Comment:
As discussed, the DRILL-5254 changes need to go in as Calcite changes. So this change can be done only after we get these changes into Calcite.